pkg/collector/corechecks/snmp/session.go

```diff
@@ -0,0 +1,125 @@
+package snmp
+
+import (
+	"fmt"
+	"github.com/gosnmp/gosnmp"
+	"time"
+)
+
+const sysObjectIDOid = "1.3.6.1.2.1.1.2.0"
+
+type sessionAPI interface {
+	Configure(config snmpConfig) error
+	Connect() error
+	Close() error
+	Get(oids []string) (result *gosnmp.SnmpPacket, err error)
+	GetBulk(oids []string) (result *gosnmp.SnmpPacket, err error)
+	GetNext(oids []string) (result *gosnmp.SnmpPacket, err error)
+	GetVersion() gosnmp.SnmpVersion
+}
+
+type snmpSession struct {
+	gosnmpInst gosnmp.GoSNMP
+}
+
+func (s *snmpSession) Configure(config snmpConfig) error {
+	if config.oidBatchSize > gosnmp.MaxOids {
+		return fmt.Errorf("config oidBatchSize (%d) cannot be higher than gosnmp.MaxOids: %d", config.oidBatchSize, gosnmp.MaxOids)
+	}
+
+	if config.communityString != "" {
+		if config.snmpVersion == "1" {
+			s.gosnmpInst.Version = gosnmp.Version1
+		} else {
+			s.gosnmpInst.Version = gosnmp.Version2c
+		}
+		s.gosnmpInst.Community = config.communityString
+	} else if config.user != "" {
+		authProtocol, err := getAuthProtocol(config.authProtocol)
+		if err != nil {
+			return err
+		}
+
+		privProtocol, err := getPrivProtocol(config.privProtocol)
+		if err != nil {
+			return err
+		}
+
+		msgFlags := gosnmp.NoAuthNoPriv
+		if privProtocol != gosnmp.NoPriv {
+			// Auth is needed if privacy is used.
+			// "The User-based Security Model also prescribes that a message needs to be authenticated if privacy is in use."
+			// https://tools.ietf.org/html/rfc3414#section-1.4.3
+			msgFlags = gosnmp.AuthPriv
+		} else if authProtocol != gosnmp.NoAuth {
+			msgFlags = gosnmp.AuthNoPriv
+		}
+
+		s.gosnmpInst.Version = gosnmp.Version3
+		s.gosnmpInst.MsgFlags = msgFlags
+		s.gosnmpInst.ContextName = config.contextName
+		s.gosnmpInst.SecurityModel = gosnmp.UserSecurityModel
+		s.gosnmpInst.SecurityParameters = &gosnmp.UsmSecurityParameters{
+			UserName:                 config.user,
+			AuthenticationProtocol:   authProtocol,
+			AuthenticationPassphrase: config.authKey,
+			PrivacyProtocol:          privProtocol,
+			PrivacyPassphrase:        config.privKey,
+		}
+	} else {
+		return fmt.Errorf("an authentication method needs to be provided")
+	}
+
+	s.gosnmpInst.Target = config.ipAddress
+	s.gosnmpInst.Port = config.port
+	s.gosnmpInst.Timeout = time.Duration(config.timeout) * time.Second
+	s.gosnmpInst.Retries = config.retries
+
+	// Uncomment following line for debugging
+	// s.gosnmpInst.Logger = log.New(os.Stdout, "", 0)
+	return nil
+}
+
+func (s *snmpSession) Connect() error {
+	return s.gosnmpInst.Connect()
+}
+
+func (s *snmpSession) Close() error {
+	return s.gosnmpInst.Conn.Close()
+}
+
+func (s *snmpSession) Get(oids []string) (result *gosnmp.SnmpPacket, err error) {
+	return s.gosnmpInst.Get(oids)
+}
+
+func (s *snmpSession) GetBulk(oids []string) (result *gosnmp.SnmpPacket, err error) {
+	return s.gosnmpInst.GetBulk(oids, 0, 10)
```

Comment:
nit: move `10` into a `BULK_SIZE` const

Reply:
Updated here to use a const: https://github.com/DataDog/datadog-agent/commit/c582e592c970ca35d946a918f04896e97b55d655#diff-6967cb5266a6aad9a0fd2f311f34b2c474352ab8d74675c354e3d1911ff47fbaR12-R14